jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DosHandler.java

```diff
@@ -0,0 +1,559 @@
+         */
+        public StatusRejectHandler(int status)
+        {
+            _status = status >= 0 ? status : HttpStatus.TOO_MANY_REQUESTS_429;
+        }
+
+        @Override
+        public boolean handle(Request request, Response response, Callback callback) throws Exception
+        {
+            if (_status == 0)
+                callback.failed(new RejectedExecutionException());
+            else
+                Response.writeError(request, response, callback, _status);
+            return true;
+        }
+    }
+
+    /**
+     * A Handler to reject DoS requests after first delaying them.
+     */
+    public static class DelayedRejectHandler extends Handler.Abstract
+    {
+        record Exchange(Request request, Response response, Callback callback)
+        {}
+
+        private final AutoLock _lock = new AutoLock();
+        private final Deque<Exchange> _delayQueue = new ArrayDeque<>();
+        private final int _maxDelayQueue;
+        private final long _delayMs;
+        private final Request.Handler _reject;
+        private Scheduler _scheduler;
+
+        public DelayedRejectHandler()
+        {
+            this(-1, -1, null);
+        }
+
+        /**
+         * @param delayMs The delay in milliseconds to hold rejected requests before sending a response or -1 for a default (1000ms)
+         * @param maxDelayQueue The maximum number of delayed requests to hold or -1 for a default (1000ms).
+         * @param reject The {@link Request.Handler} used to reject {@link Request}s or null for a default ({@link HttpStatus#TOO_MANY_REQUESTS_429}).
+         */
+        public DelayedRejectHandler(
+            @Name("delayMs") long delayMs,
+            @Name("maxDelayQueue") int maxDelayQueue,
+            @Name("reject") Request.Handler reject)
+        {
+            _delayMs = delayMs >= 0 ? delayMs : 1000;
+            _maxDelayQueue = maxDelayQueue >= 0 ? maxDelayQueue : 1000;
+            _reject = Objects.requireNonNullElseGet(reject, () -> new StatusRejectHandler(HttpStatus.TOO_MANY_REQUESTS_429));
+        }
+
+        @Override
+        protected void doStart() throws Exception
+        {
+            super.doStart();
+            _scheduler = getServer().getScheduler();
+            addBean(_scheduler);
+        }
+
+        @Override
+        protected void doStop() throws Exception
+        {
+            super.doStop();
+            removeBean(_scheduler);
+            _scheduler = null;
+        }
+
+        @Override
+        public boolean handle(Request request, Response response, Callback callback) throws Exception
+        {
+            try (AutoLock ignored = _lock.lock())
+            {
+                while (_delayQueue.size() >= _maxDelayQueue)
+                {
+                    Exchange exchange = _delayQueue.removeFirst();
+                    if (!_reject.handle(exchange.request, exchange.response, exchange.callback))
```

Comment:
I don't think we can call user code with a lock held.

Reply:
ooooops